### PR TITLE
Add PCLI command to sign account balance files

### LINKED ISSUE #5820: Add PCLI command to sign account balance files
- Some functionality exists in the platform `FileSignTool` that purports to do this, but it's possible the format has changed
- After implementing this feature, `FileSignTool` can be removed, since all the functionality will have been moved

### PR DESCRIPTION
**Pull request by iwsimon**
**Description**:

**Related issue(s)**:

Fixes #5820

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: settings.gradle.kts
Comment (povolev15) [inline settings.gradle.kts:147]: I am not sure that we can use this version in development. we need first to merge the platform into there devlop and regenerate the version from devlop and than update it.

**Review by povolev15 — APPROVED** [review]

File: settings.gradle.kts
Comment (iwsimon) [reply settings.gradle.kts:147]: Yes, I knew. Not sure what to put there yet. Wait for platform new build to put there.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (JeffreyDallas) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:101]: Why not use logger?

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (JeffreyDallas) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:97]: Why not use logger?

File: settings.gradle.kts
Comment (povolev15) [reply settings.gradle.kts:147]: maybe let's merge without it and then we can change this tag and update it on develop in services.

**Review by povolev15 — DISMISSED** [review]

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:97]: Is that really the way you want to use a cli tool?  All my CLI stuff I did with `System.out` and only left logs on for `ERROR` (and higher) in case the signed state file stuff had problems.  As a tool _user_ the result was more what I wanted to see.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (david-bakin-sl) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:104]: Why throw here instead of `System.err` like in the previous catch clause?   (Especially since on a quick look this will fly off the top of the stack - anyway it isn't handled by caller `SignCommand.signFile`.)

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (david-bakin-sl) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:115]: Not strictly necessary since `BufferedOutputStream` will flush when auto-closed.  (Actually, it's a superclass that'll do it.)

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (david-bakin-sl) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java:39]: What does this comment mean?  The constant it is attached to is written as a 4byte integer in `generateSigBalanceFile`.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (david-bakin-sl) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:117]: I guess use of `System.{out,err}` vs logging should be consistent.

**Review by david-bakin-sl — COMMENTED** [review]
Some consistency between `System.{out|err}` vs logging vs throwing an exception never handled would be an improvement (especially as this tool will be used in a production environment by professionals (who aren't developers).

Otherwise just minor comments.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:97]: This is used for command line tool. I tried to use logger and didn't see the logged msg. Talked to @alittley. Platform uses System.out (err) too.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:101]: same reason below

File: settings.gradle.kts
Comment (iwsimon) [reply settings.gradle.kts:147]: Without change, code won't compile,

Comment (iwsimon) [reply settings.gradle.kts:147]: this is adhoc platform build#.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java:51]: updated.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/BalanceFileSignCommand.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/BalanceFileSignCommand.java:32]: yes, it's nice to extend it.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:104]: Follow the platform's code style. Since it's RuntimeException, didn't have to catch it. @alittley

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java:39]: This will be used for record stream file, not account balance file yet.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:117]: yes, will update all. Thanks for catching.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:115]: yea, just to be sure here.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java:65]: updated. thanks.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java:67]: aggreed. kept as it was from the copied code.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java:53]: updated

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:98]: Yes but the question is: what happens to it?  Does it fly off the top of the stack and then the user sees a stack trace with some message buried somewhere in it?  Or is it handled by some caller and changed into a `System.err.println` that the user can understand?

**Review by david-bakin-sl — DISMISSED** [review]
LGTM.  One final comment that you can address or not as you choose.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:98]: You are right. We should not just throw exception. I will add System.err as I am working on the record stream sign.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (alittley) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:58]: Did you test whether this constructable registry setup is actually necessary? If not, this method could be removed

In any case, the commented out block below can be removed

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/ServiceSignCommand.java
Comment (alittley) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/ServiceSignCommand.java:28]: The command structure chosen here works, but it's inverted from the way existing platform commands are laid out:

Existing platform command structure:
`pcli <object> <operation> <...>` -> `pcli event-stream sign <...>`

Command structure in this PR:
`pcli <operation> <object> <...>` -> `pcli service-sign balance <...>`

The following changes would bring your structure in line with existing commands:
- Instead of `ServiceSignCommand`, you would have an `AccountBalanceCommand` class, which extends `AbstractCommand`, is tagged `@SubcommandOf(PlatformCli.class)`, with `name = account-balance`
- The `BalanceFileSignCommand` would be tagged `@SubcommandOf(AccountBalanceCommand)`, extend `SignCommand`, with `name = sign`

Therefore, the command to use the tool would be
```
pcli account-balance sign <...> 
```

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (alittley) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java:48]: This functionality is implemented in `FileSigningUtils::signData`. Is there a reason to reimplement here?

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java
Comment (alittley) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java:25]: Are account balance files considered to be stream files?

In any case, it seems this class is only being used to determine if a given file has the correct file extension. Unless there are plans to further utilize this class in the future, I would suggest simply doing the check manually in `BalanceFileSignCommand::isFileSupported`

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java]: tested it no problem without this, removed.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/ServiceSignCommand.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/ServiceSignCommand.java:28]: Services have two types of stream files to sign at this time, pcli serive-sign balance and record-stream. Maybe more later. So I setup service-sign at the top level.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java:25]: I might use some other functions for testing later. Kept it for now. Just in a rush to get this in for tomorrow's build.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java]: Updated. Thanks

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/ServiceSignCommand.java
Comment (alittley) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/ServiceSignCommand.java:28]: The idea is that all signing methods share the same command, `sign`. This is fully compatible with having multiple types of stream files to sign.

The current structure is not consistent with the intended usage

**Review by david-bakin-sl — APPROVED** [review]
LGTM

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/ServiceSignCommand.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/ServiceSignCommand.java]: I updated the code. Now you can run it as : pcli account-balance sign XXX

**Review by povolev15 — DISMISSED** [review]

**Review by jeromy-cannon — APPROVED** [review]

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceCommand.java
Comment (alittley) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceCommand.java:25]: class description is out of date

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java
Comment (alittley) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java:25]: So are account balance files considered stream files? The fact that multiple required overrides are not applicable makes me think they should not be.

If that is the case, it seems counter productive to make an `AccountBalanceType` which inherits from `StreamType`

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceCommand.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceCommand.java]: Updated. Thanks!

**Review by povolev15 — APPROVED** [review]

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceCommand.java
Comment (Neeharika-Sompalli) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceCommand.java:26]: Probably will help to describe what all collection of operations can be done.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java
Comment (alittley) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java:25]: But if it isn't a stream file in the first place, then `StreamType.isStreamFile()` doesn't make any sense

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSignCommand.java
Comment (Neeharika-Sompalli) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSignCommand.java:47]: `isStreamFile` should only be used for `EVENT_STREAM` and `RECORD_STREAM`

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (Neeharika-Sompalli) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java:25]: Why are we using `TYPE_*` here ?

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java
Comment (Neeharika-Sompalli) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java:25]: Agreed with Austin

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java
Comment (povolev15) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/SignUtils.java:25]: this type will be used for record stream

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceCommand.java
Comment (povolev15) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceCommand.java:26]: we will add it is just sign as of right now

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSignCommand.java
Comment (povolev15) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSignCommand.java:47]: we are using this function to compare the extension and if it is supported

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java
Comment (povolev15) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java:25]: @alittley 
StreamType.isStreamFile() is used for file comparison in all other places what kind of file we are receiving. we are comparing the

Comment (alittley) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java:25]: It's being used in cases where the type of file being searched for *is a stream file*.

Why not just check the extension directly, rather than inheriting that functionality from a parent class which doesn't logically make sense?

**Review by povolev15 — APPROVED** [review]

**Review by jeromy-cannon — DISMISSED** [review]
from a release engineering perspective I approve the build changes, please wait for other approvers for remaining code before merging.

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:81]: Seems to not be changed yet.  Remember it for next time.

**Review by david-bakin-sl — APPROVED** [review]
LGTM though I still don't know whether the `swirlds-version` version hash is the right one or not.  (As far as I'm concerned some things can be cleaned up in a subsequent PR.)

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java
Comment (povolev15) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:81]: I fixed it and return false no one handle the throw that's why return false will be good.

Comment (povolev15) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceSigningUtils.java:81]: @iwsimon followed the pattern @alittley is using an in case this exception happens it is more serious and we want to throw an exception with all the stuck traces. Any issue with that?

**Review by david-bakin-sl — DISMISSED** [review]
LGTM again ....

**Review by povolev15 — APPROVED** [review]

**Review by jeromy-cannon — APPROVED** [review]

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/AccountBalanceType.java]: removed StreamType from AccountBalanceType

**Review by david-bakin-sl — DISMISSED** [review]
The is a command line tool.  For production use yes.  But a CLI, not a server we're depending on.  I think it is ok for it to be rough around the edges and get cleaned up in a very closely-following PR.  Am I alone in this?

**Review by povolev15 — APPROVED** [review]

**Comment by povolev15** [comment]
> The is a command line tool. For production use yes. But a CLI, not a server we're depending on. I think it is ok for it to be rough around the edges and get cleaned up in a very closely-following PR. Am I alone in this?

I agree with you but this tool was decided by Nathan as production on the channel and we need to merge it today, it is an endless story let's try to focus on merging it

File: hedera-node/cli-clients/src/main/java/module-info.java
Comment (qnswirlds) [inline hedera-node/cli-clients/src/main/java/module-info.java:13]:
```suggestion
```
As you don't use logging, these lines shouldn't be here.

**Review by qnswirlds — APPROVED** [review]

File: hedera-node/cli-clients/src/main/java/module-info.java
Comment (iwsimon) [reply hedera-node/cli-clients/src/main/java/module-info.java]: Removed it in the next PR for record stream sign. Thanks.

**Review by povolev15 — APPROVED** [review]

**Comment by nickpoorman** [comment]
@david-bakin-sl, what are the things that are rough around the edges that you would like fixed? We should get those done today before merging this in.

**Comment by nickpoorman** [comment]
@iwsimon and @povolev15, this also needs tests before it can be merged.

**Review by JeffreyDallas — APPROVED** [review]

**Comment by david-bakin-sl** [comment]
> @david-bakin-sl, what are the things that are rough around the edges that you would like fixed? We should get those done today before merging this in.

As far as  I'm concerned there's _nothing_ "rough around the edges" that I would like fixed.  I've done the "LGTM" thing about 10 times already on this review.  Which is a PR that we're all waiting for.  But changes are being consistently suggested for this _first release of a command line tool_.  I was trying to suggest (apparently badly) that we just _get it in now_ and fix it when the pressure to get a release tagged is off.  Because, even if it will be used in production immediately: _command line tool_.  Yes it may blow up (and emit a stackdump).  Yes the command order might not match what's already there.  And so on.  But _command line tool_ not 24/7/365 6-nines service.  That's all I meant.

**Review by david-bakin-sl — APPROVED** [review]
LGTM _one more time for luck!_

**Pull request merged**
